File: java/factory/FactoryDemo.java
//author sbkun
package factory;

abstract class HwPhone {
    HwPhone(String model) {
        this.model = model;
    }

    void call(String to) {
        if (!model.isEmpty()) {
            System.out.println(model + " is calling " + to);
        }
    }

    private String model;

    //手机型号
    public static final String LawMate30 = "LowMate30";
    public static final String LowHonour6 = "LowHonour6";
    public static final String LowHonour7 = "LowHonour7";

    public static final String HighMate30 = "HighMate30";
    public static final String HighHonour6 = "HighHonour6";
    public static final String HighHonour7 = "HighHonour7";
}

class LowMate30 extends HwPhone {
    LowMate30() {
        super(HwPhone.LawMate30);
    }
}

class LowHonour6 extends HwPhone {
    LowHonour6() {
        super(HwPhone.LowHonour6);
    }
}

class LowHonour7 extends HwPhone {
    LowHonour7() {
        super(HwPhone.LowHonour7);
    }
}

class SimpleFactory {
    public static HwPhone makePhone(String model) {
        if (model.equalsIgnoreCase(HwPhone.LawMate30)) {
            return new LowMate30();
        } else if (model.equalsIgnoreCase(HwPhone.LowHonour6)) {
            return new LowHonour6();
        } else if (model.equalsIgnoreCase(HwPhone.LowHonour7)) {
            return new LowHonour7();
        }
        return null;
    }
}

//工厂方法
interface PhoneFactory {
    HwPhone makePhone();
}

class LowMate30Factory implements PhoneFactory {
    @Override
    public HwPhone makePhone() {
        return new LowMate30();
    }
}

class LowHonour6Factory implements PhoneFactory {
    @Override
    public HwPhone makePhone() {
        return new LowHonour6();
    }
}

//抽象工厂
interface HonourFactory {
    //抽象工厂，负责创建荣耀系列手机
    HwPhone makeHonour6();

    HwPhone makeHonour7();
}

//具体工厂类，负责创建低配的，荣耀系列手机
class LowHonourFactory implements HonourFactory {
    @Override
    public HwPhone makeHonour6() {
        return new LowHonour6();
    }

    @Override
    public HwPhone makeHonour7() {
        return new LowHonour7();
    }
}

public class FactoryDemo {
    static void testSimpleFactory() {
        System.out.println("testSimpleFactory");
        HwPhone p1 = SimpleFactory.makePhone(HwPhone.LawMate30);
        p1.call("110");
        HwPhone p2 = SimpleFactory.makePhone(HwPhone.LowHonour6);
        p2.call("110");
    }

    static void testFactoryMethod() {
        System.out.println("testFactoryMethod");
        PhoneFactory f1 = new LowMate30Factory();
        HwPhone p1 = f1.makePhone();
        p1.call("110");

        PhoneFactory f2 = new LowHonour6Factory();
        HwPhone p2 = f2.makePhone();
        p2.call("110");
    }

    static void testAbstractFactory() {
        System.out.println("testAbstractFactory");
        HonourFactory f = new LowHonourFactory();
        HwPhone p1 = f.makeHonour6();
        HwPhone p2 = f.makeHonour7();
        p1.call("110");
        p2.call("110");
    }

    public static void main(String args[]) {
        testSimpleFactory();
        System.out.println("\n====================\n");
        testFactoryMethod();
        System.out.println("\n====================\n");
        testAbstractFactory();
    }
}

File: cpp/creational/FactoryDemo.cpp
//author sbkun
#include <iostream>
#include <string>
#include <string.h>

using namespace std;

//华为手机抽象类
class HwPhone {
	public:
		explicit HwPhone(const string& model): model_(model) {}
		virtual ~HwPhone() = default;

		virtual void call(const string& to) {
			if (!model_.empty()){
				cout<<model_<<" is calling "<<to<<endl;
			}
		}
	protected:
	//具体型号
		string model_;

	public:
	//低配版 Mate30 Honour6 Honour7
 	static constexpr const char* LowMate30  = "LowMate30";
	static constexpr const char* LowHonour6 = "LowHonour6";
	static constexpr const char* LowHonour7 = "LowHonour7";

	//高配版 Mate30 Honour6 Honour7
 	static constexpr const char* HighMate30  = "HighMate30";
	static constexpr const char* HighHonour6 = "HighHonour6";
	static constexpr const char* HighHonour7 = "HighHonour7";
};

//Huawe mate30 手机
class LowMate30: public HwPhone{
	public:
		LowMate30() : HwPhone(HwPhone::LowMate30) {}		
};

//Huawe Honour6 手机
class LowHonour6: public HwPhone{
	public:
		LowHonour6() : HwPhone(HwPhone::LowHonour6) {}		
};

//Huawe Honour7 手机
class LowHonour7: public HwPhone{
	public:
		LowHonour7() : HwPhone(HwPhone::LowHonour7) {}		
};

//简单工厂: 一个方法负责创建所有产品
class SimpleFactory {
	public:
		static HwPhone* makePhone(const string& model) {
			if (strcasecmp(model.c_str(), HwPhone::LowMate30) == 0) {
				return new LowMate30;
			}
			if (strcasecmp(model.c_str(), HwPhone::LowHonour6) == 0) {
				return new LowHonour6;
			}
			if (strcasecmp(model.c_str(), HwPhone::LowHonour7) == 0) {
				return new LowHonour7;
			}
			return NULL;
		}
};

//工厂方法: 一个工厂负责创建一个产品
class PhoneFactory {
	public:
		virtual ~PhoneFactory() = default;
		virtual HwPhone* makePhone() = 0;
};

class LowMate30Factory: public PhoneFactory{
	public:
		HwPhone* makePhone() override{
			return new LowMate30;
		}
};

class LowHonour6Factory: public PhoneFactory {
	public:
		virtual HwPhone* makePhone() override {
			return new LowHonour6;
		}
};

//抽象工厂： 一个工厂创建一个产品族
class HonourFactory {
	//抽象工厂，负责创建荣耀系列手机
	public:
	virtual ~HonourFactory() = default;
	virtual HwPhone* makeHonour6() = 0; 
	virtual HwPhone* makeHonour7() = 0;
};

//具体工厂类，负责创建低配的，荣耀系列手机
class LowHonourFactory: public HonourFactory{
	public:
	HwPhone* makeHonour6() override {
		return new LowHonour6;		
	}

	HwPhone* makeHonour7() override {
		return new LowHonour7;
	}
};

void testSimleFactory() {
	cout<<__func__<<endl<<endl;

	HwPhone* p1 = SimpleFactory::makePhone(HwPhone::LowMate30);
	p1->call("110");
	delete p1;

	HwPhone* p2 = SimpleFactory::makePhone(HwPhone::LowHonour6);
	p2->call("110");
	delete p2;

	cout<<endl<<"========="<<endl<<endl;
}

void testFactoryMethod(){
	cout<<__func__<<endl<<endl;

	PhoneFactory* f1 = new LowMate30Factory;
	HwPhone* p1 = f1->makePhone();
	p1->call("110")	;
	delete f1;
	delete p1;
	
	PhoneFactory* f2 = new LowHonour6Factory;
	HwPhone* p2 = f2->makePhone();
	p2->call("110")	;
	delete f2;
	delete p2;

	cout<<endl<<"========="<<endl<<endl;
}

void testAbstractFactory() {
	cout<<__func__<<endl<<endl;

	HonourFactory* f1 = new LowHonourFactory;
	HwPhone* p1 = f1->makeHonour6();
	p1->call("110")	;
	
	HwPhone* p2 = f1->makeHonour7();
	p2->call("110")	;
	delete f1;
	delete p1;
	delete p2;
}

int main(int argc, char* argv[])
{
	testSimleFactory();
	testFactoryMethod();
	testAbstractFactory();
	return 0;
}
